test/e2e/upgrade_cluster_test.go

```diff
@@ -0,0 +1,189 @@
+//go:build e2e
+// +build e2e
+
+package e2e
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"testing"
+
+	"github.com/onsi/gomega"
+	hyperv1 "github.com/openshift/hypershift/api/hypershift/v1beta1"
+	e2eutil "github.com/openshift/hypershift/test/e2e/util"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	nodePoolAnnotationCurrentConfig        = "hypershift.openshift.io/nodePoolCurrentConfig"
+	nodePoolAnnotationCurrentConfigVersion = "hypershift.openshift.io/nodePoolCurrentConfigVersion"
+	hostedClusterUpgradeTestLabel          = "hypershift.openshift.io/upgrade-test"
+)
+
+// TestUpgradeHyperShift validates that a HyperShift Operator upgrade won't unnecessarily cause a node rollout.
+// The test must be triggered via the flag upgrade.run-tests and the HyperShift Operator initial image must be set
+// via the flag e2e.hypershift-operator-initial-image.
+func TestUpgradeHyperShift(t *testing.T) {
+	if globalOpts.RunUpgradeTest {
+		var once sync.Once
+		var cleanup e2eutil.Cleanup
+		var err error
+
+		var mgmtClient crclient.Client
+		var hostedCluster *hyperv1.HostedCluster
+
+		var nodePoolsMap map[string]*hyperv1.NodePool
+
+		hyperShiftOperatorLatestImage := globalOpts.HyperShiftOperatorLatestImage
+		hyperShiftOperatorInitialImage := globalOpts.HyperShiftOperatorInitialImage
+
+		t.Parallel()
+		ctx, cancel := context.WithCancel(testContext)
+		defer cancel()
+
+		clusterOpts := globalOpts.DefaultClusterOptions(t)
+
+		t.Log("Starting HyperShift Operator upgrade test")
+		zones := strings.Split(globalOpts.ConfigurableClusterOptions.Zone.String(), ",")
+		if len(zones) >= 3 {
+			t.Logf("Sufficient zones available for InfrastructureAvailabilityPolicy HighlyAvailable")
+			clusterOpts.AWSPlatform.Zones = zones
+			clusterOpts.InfrastructureAvailabilityPolicy = string(hyperv1.HighlyAvailable)
+			clusterOpts.NodePoolReplicas = 1
+		}
+
+		clusterOpts.BeforeApply = func(o crclient.Object) {
+			switch obj := o.(type) {
+			case *hyperv1.HostedCluster:
+				if obj.Annotations == nil {
+					obj.Annotations = make(map[string]string)
+				}
+				obj.Annotations[hyperv1.ControlPlaneOperatorV2Annotation] = "true"
+
+				// Add a label to identify the HostedCluster for upgrade tests in case they're leaked
+				// and need to be cleaned up.
+				if obj.Labels == nil {
+					obj.Labels = make(map[string]string)
+				}
+				obj.Labels[hostedClusterUpgradeTestLabel] = "true"
+			}
+		}
+
+		t.Run("Install initial HyperShift Operator", func(tt *testing.T) {
+			tt.Logf("Installing HyperShift Operator with image %s", hyperShiftOperatorInitialImage)
+			globalOpts.HyperShiftOperatorLatestImage = hyperShiftOperatorInitialImage
+			cleanup, err = e2eutil.InstallHyperShiftOperatorFull(tt, ctx, globalOpts)
+			if err != nil {
+				tt.Error("Failed to install initial HyperShift Operator")
+				tt.Fail()
+				t.FailNow()
```

Comment:
you can't do that, `FailNow()` must be called from the same goroutine. `t.Run()` starts a new goroutine.

It's also not needed, if a subtest fails, the outer test also fail.

Reply:
Ah yeah I totally forgot about the same goroutine.

> It's also not needed, if a subtest fails, the outer test also fail.

I believe it is needed as when the subtest fails, the outer test doesn't fail immediately, instead Run returns false. It's why on other Run() executions I used the returned value to set the outer test as failed. Here's the doc for Run

```
// Run runs f as a subtest of t called name. It runs f in a separate goroutine
// and blocks until f returns or calls t.Parallel to become a parallel test.
// Run reports whether f succeeded (or at least did not fail before calling t.Parallel).
//
// Run may be called simultaneously from multiple goroutines, but all such calls
// must return before the outer test function for t returns.
```

Reply:
right, if the point is to stop execution afterwards, they yeah checking the return value of `t.Run()` or checking `t.Failed()` is the correct way to do it.

Reply:
Addressed